src/psyclone/dynamo0p3.py

```diff
@@ -7452,137 +7451,136 @@ def gen_code(self, parent):
-#        self._parent.add(CommentGen(self._parent, ""),
-#                         position=["before", self._position])
-#        ArgOrdering.generate(self)
-#
-#    def _add_dino_scalar(self, name):
-#        ''' add a dino output call for a scalar variable '''
-#        self._parent.add(CallGen(self._parent, name="dino%output_scalar",
-#                                 args=[name]),
-#                         position=["after", self._scalar_position])
-#
-#    def _add_dino_array(self, name):
-#        ''' add a dino output call for an array variable '''
-#        self._parent.add(CallGen(self._parent, name="dino%output_array",
-#                                 args=[name]),
-#                         position=["after", self._array_position])
+    def get_kernel_schedule(self):
+        '''Returns an LFRic PSyIR Schedule representing the kernel code. The
+        base class creates the PSyIR schedule on first invocation
+        which is then transformed to an LFRic PSyIR Schedule here. The
+        Schedule is just generated on first invocation, this allows us
+        to retain transformations that may subsequently be applied to
+        the Schedule (but will not adapt to transformations applied to
+        the fparser2 AST).
+
+        :returns: LFRic Schedule representing the kernel code.
+        :rtype: :py:class:`psyclone.psyGen.KernelSchedule`
+
+        '''
+        if self._kern_schedule is None:
+            # Get the PSyIR Kernel Schedule
+            psyir_schedule = super(DynKern, self).get_kernel_schedule()
+
+            # Before transforming, check the validity of the kernel
+            # arguments
+            self.validate_kernel_code_args()
+
+            # TODO replace the PSyIR argument data symbols with LFRic
+            # data symbols, see issue #XXX. For the moment we simply
+            # return the unmodified PSyIR schedule
+            self._kern_schedule = psyir_schedule
+
+        return self._kern_schedule
+
+    def validate_kernel_code_args(self):
+        '''Check that the arguments in the kernel code match the expected
+        arguments as defined by the kernel metadata and the LFRic
+        API.
+
+        '''
+        # Get the kernel code arguments
+        psyir_schedule = super(DynKern, self).get_kernel_schedule()
+        symbol_table = psyir_schedule.symbol_table
+        kern_code_args = symbol_table.argument_list
+
+        # Get the kernel code interface according to the kernel
+        # metadata and LFRic API
+        interface_info = KernelInterface(self)
+        interface_info.generate()
+        interface_args = interface_info.arglist
+
+        # 1: Check the the number of arguments match
+        actual_n_args = len(kern_code_args)
+        expected_n_args = len(interface_args)
+        if actual_n_args != expected_n_args:
+            raise GenerationError(
+                "The number of arguments expected by the psy-layer kernel "
```

Comment:
Please say something about the implemented kernel interface versus that specified in meta-data - I think that will be clearer.

Reply:
I've reworded. I think it sounds better. Hopefully it is along the lines of what you are after.